### Generator 函数
- Generator 函数是 es6 提供的一种异步编程解决方案,语法行为与传统函数完全不同
  - 1.Generator 函数是一个状态机,封装了多个内部状态
  - 2.执行 Gennerator 函数会返回一个遍历器对象(Iteraotr Object),可以依次遍历 Generator 函数内部的每个状态
- Generator 函数是个普通函数,但有两个特征:
  - 1.function 关键字与函数名之间有一个星号
  - 2.函数体内部使用yield表达式,定义不同的内部状态(yield:"产出")
  - *调用后,函数并不执行,返回的也不是函数运行结果,而是一个指向内部状态的指针对象(Iterator)*
  ```js
  function* gen1() {
    console.log(1); yield 'hello';
    console.log(2); yield 'world';
    console.log(3); return 'ok';
  }
  const iter = gen1();
  iter.next(); // 1 {value: "hello", done: false}
  iter.next(); // 2 {value: "world", done: false}
  iter.next(); // 3 {value: "ok", done: true}
  iter.next(); // {value: undefined, done: true}
  ```

#### yield 表达式
- 提供了一种可以暂停执行的函数,yield 表达式就是暂停标志
- yield 表达式后面的表达式,只有当调用 next 方法,内部指针指向该语句时才会执行
  - 相当于提供了手动的"惰性"求值的语法功能
  ```js
  function* f1() { console.log('执行内容'); }
  const iter = f1();
  setTimeout(() => iter.next(), 1000);
  ```

#### 与 Iteraltro 接口的关系
- 由于 Generator 函数就是遍历器生成函数
  - 因此可以把 Generator 赋值给对象的 Symbol.iterator 属性
  - 从而使该对象具有 Iterator 接口
  ```js
  const obj = {
    [Symbol.iterator]: function* () {
      yield 'a';
      yield* [...'bc'];
    }
  };
  [...obj]; // ["a", "b", "c"]
  ```

#### next 方法的参数
- yield 表达式本身没有返回值,或者说总是返回 undefined
- next 方法可以带一个参数,该参数会被当作*上一个yield表达式的返回值*
  ```js
  function* f() {
    for (let i = 0; true; i++) {
      let reset = yield i;
      if (reset) i = -1;
    }
  }
  const g = f();
  g.next(); // {value: 0, done: false}
  g.next(); // {value: 1, done: false}
  g.next(true); // {value: 0, done: false}
  ```

#### for...of循环
- for...of 循环可以自动遍历 Generator 函数运行时生成的 Iterator 对象,且此时不再需要调用 next 方法
  - *一旦next方法的返回对象的done属性为true,循环就会中止,且不包含该返回对象*
  ```js
  function* foo() {
    yield 1;
    yield 2;
    return 3;
  }
  for (let v of foo()) console.log(v); // 1 2
  [...foo()] // 1 2
  ```
- 斐波那契额数列
  ```js
  function* f() {
    let a = 0, b = 1;
    while(true) {
      yield a;
      [a, b] = [b, a + b];
    }
  }
  for (let v of f()) {
    if (v > 1000) break;
    console.log(v);
  }
  ```

#### Generator.prototype.throw()
- Generator 函数返回的遍历器对象,都有一个 throw 方法,可以在函数体内抛出错误,然后在 Generator 函数体内捕获
  - 第一个错误被 Generator 函数体内的catch 语句捕获
  - 第二次抛出错误,Generator 函数内部的catch语句执行过了,不会再捕获这个错误了
  - throw 方法抛出的错误要被内部捕获,前提是至少执行过一次 next 方法
  - throw方法被捕获后,自动执行一次 next 方法
  - 一旦 Generator 抛出错误,且没有被内部捕获,就不会在执行下去,如果还调用 next 方法,将返回 {done: true}
  ```js
  function* g() {
    try {
      yield;
    } catch (e) {
      console.log(e);
    }
  }
  const iter = g();
  iter.next();
  iter.throw('a'); // a
  iter.throw('b'); // VM1491:11 Uncaught b
  ```

#### Generator.prototype.return()
- return 方法,可以返回给定值,并终结遍历 Generator 函数
  ```js
  function* g() {
    yield 1;
    yield 2;
    yield 3;
  }
  const iter = g();
  iter.next(); // {value: 1, done: false}
  iter.return('hello'); // {value: "hello", done: true}
  iter.next(); // {value: undefined, done: true}
  ```
- 如果 Gnerator 函数内部有 try...finally 代码块,且正在执行 try 代码块
  - return 方法会导致立即进入 finally 代码块,执行结束后,整个函数才会结束
  ```js
  function* g() {
    yield 1;
    try {
      yield 2;
      yield 3;
    } finally {
      yield 4;
      yield 5;
    }
    yield 6;
  }
  const iter = g();
  iter.next(); // {value: 1, done: false}
  iter.next(); // {value: 2, done: false}
  iter.return('ok'); // {value: 4, done: false}
  iter.next(); // {value: 5, done: false}
  iter.next(); // {value: undefined, done: true}
  ```

#### next()、throw()、return() 的共同点
- 三个方法本质上是同一件事,作用都是让 Generator 函数恢复执行,并且使用不同的语句替换 yield 表达式
  ```js
  function* g() {
    let i = 1;
    i = yield i;
    yield i;
  }
  const iter = g();
  iter.next();
  iter.next(2);
  ```
- 1.next() 将 yield 表达式替换成一个值
  ```js
  iter.next(1);
  // 相当于
  let result = yield x + y;
  // 换成
  let result = 1;
  ```
- 2.throw() 将 yield 表达式替换成一个 throw 语句
  ```js
  iter.throw(new Error('出错了'));
  // 相当于
  let result = yield x + y;
  // 换成
  let result = throw(new Error('出错了'));
  ```
- 3.return() 将 yield 表达式替换成一个 return 语句
  ```js
  iter.return(2);
  // 相当于
  let result = yield x + y;
  // 换成
  let result = return 2;
  ```

#### yield* 表达式
- 如果在 Generator 函数内部,调用另一个 Generator 函数,需要在前者的函数体内,手动完成遍历
  ```js
  function* g1() {
    yield 1;
    yield 2;
  }
  function* g2() {
    for (let i of g1()) {
      yield i;
    }
    yield 3;
    yield 4;
  }
  [...g2()] // 1 2 3 4
  ```
- es6提供了 yield* 表达式,用来在一个 Generator 函数内执行另一个 Generator 函数
  ```js
  function* g2() {
    yield* g1();
    yield 3;
    yield 4;
  }
  ```
- 任何数据结构,只要有 Iterator 接口,就可以被 yield* 遍历
  ```js
  function* g2() {
    yield* [1, 2];
    yield* '12';
  }
  ```

#### 作为对象属性的 Generator 函数
- 如果一个对象的属性是 Generator 函数,可以简写成下面的形式
  ```js
  let obj = {
    * g() {}
  };
  ```
- 与下面形式等价
  ```js
  let obj = {
    g: function* () {}
  };
  ```

#### Generator 函数的 this
- Generator 函数总是返回一个遍历器,es6规定这个遍历器就是 Generator 函数的实例
  - 也继承了 Generator 函数的 prototype 对象上的方法
  ```js
  function* g() {}
  g.prototype.hello = () => 'hi!';
  let obj = g();
  obj instanceof g // true
  obj.hello() // hi!
  ```
- 但如果把g当作普通的构造函数,并不会生效,应为g返回的总是遍历器对象,而不是this对象
  ```js
  function* g() {
    this.a = 1;
    this.hello = () => 'hi!';
  }
  let obj = g();
  obj.a // undefined
  obj.hello() // undefined
  ```
- 变通的法子
  ```js
  function* g() {
    this.a = 1;
    yield this.b = 2;
    yield this.c = 3;
  }
  const f = g.call(g.prototype);
  ```

#### Generator 与 状态机
```js
// es5
let status = true;
let clock = function() {
  console.log(status ? 'on' : 'off');
  status = !status;
}
// es6
let clock = *function() {
  while (true) {
    console.log('on');
    yield;
    console.log('off');
    yield;
  }
}
```

